### PR TITLE
Bug 1429030 - automatically suggest Bugzilla product and component when bug filer gets opened

### PR DESCRIPTION
**Pull request by Archaeopteryx**
Treeherder tries to extract the source path of the file (often the failing test) and looks it up in the regularly imported mapping of source files to Bugzilla components which are provided by the `Bugzilla` tasks of the automation.

To activate the import of a repository, set the `life_cycle_order` property of the repository. For now, no distinction is made between different kinds of repositories, adding repositories other than gecko (= of the mozilla-central kind) could cause conflicting information (e.g. a `README` file in the mozilla-beta and the fenix repository).

### CODE REVIEW

**Comment by Archaeopteryx** [comment]
@sarah-clements Please have a look at the patch stack when you find time. Ingestion for mozilla-esr91 will be enabled once it has been completely set up.

**Comment by sarah-clements** [comment]
> @sarah-clements Please have a look at the patch stack when you find time. Ingestion for mozilla-esr91 will be enabled once it has been completely set up.

Sorry I missed this. I'll take a look later today.

File: treeherder/webapp/api/bug_creation.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                .values('bugzilla_component__product', 'bugzilla_component__component')
+                .distinct()
+            )
+        if len(queryset) > 0:
+            return queryset[: min(5, len(queryset))]
+        queryset = (
+            FilesBugzillaMap.objects.select_related('bugzilla_component')
+            .filter(file_name=file)
+            .values('bugzilla_component__product', 'bugzilla_component__component')
+            .distinct()
```
Comment (sarah-clements) [inline treeherder/webapp/api/bug_creation.py:50]: This is a lot of queries... it seems like it could be slow when you get a certain number of rows in the table. Since you're essentially repeating the same query several times you could create a new queryset class in models.py and add it as a 'manager' to the FilesBugzillaMap. You'll see an example in the top of the models.py file. 

And/or you could use Q objects to filter with 'OR' logic in the queryset.

File: treeherder/model/models.py [hunk 2]
```diff
@@ -333,6 +334,32 @@ def search(self, search_term):
...
+    path = models.CharField(max_length=255, unique=True, db_index=True)
+    file_name = models.CharField(max_length=255, db_index=True)
+    bugzilla_component = models.ForeignKey('BugzillaComponent', on_delete=models.CASCADE)
+
+    class Meta:
+        db_table = 'file_bugzilla_component'
+        verbose_name_plural = 'files_bugzilla_components'
+
+    def __str__(self):
+        return "{0}".format(self.path)
```
Comment (sarah-clements) [inline treeherder/model/models.py:360]: How big do you think these tables will get and will there ever be a need to expire old data?

File: tests/model/test_files_bugzilla_map.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    Test for which projects the mapping of files to Bugzilla components shall be
+    imported and if the order is correct.
+    """
+    actual_projects = list(
+        Repository.objects.filter(codebase='gecko')
+        .filter(active_status='active')
+        .filter(life_cycle_order__isnull=False)
+        .values_list('name', flat=True)
+        .order_by('life_cycle_order')
```
Comment (sarah-clements) [inline tests/model/test_files_bugzilla_map.py:25]: I'm not following the purpose of `life_cyle_order` other than to apply to certain gecko repo's and rank them? What's the ranking useful for?

**Comment by sarah-clements** [comment]
> Treeherder tries to extract the source path of the file (often the failing test) and looks it up in the regularly imported mapping of source files to Bugzilla components which are provided by the `Bugzilla` tasks of the automation.
> 

The bugzilla bug is quite old and there isn't any recent conversation or explanation about the approach you took. I see you moved the logic from the UI into a separate backend class, with the idea of storing some component and paths in tables. Could you please explain why you think this is the best approach. Such as why do we need to store the paths and component info. Is the intention for the django command you've created to be run as a cron task? Will this be a fairly small table and there's no need to ever expire data, etc. 

> To activate the import of a repository, set the `life_cycle_order` property of the repository. For now, no distinction is made between different kinds of repositories, adding repositories other than gecko (= of the mozilla-central kind) could cause conflicting information (e.g. a `README` file in the mozilla-beta and the fenix repository).

Why use a numerical value rather a boolean, if its only used to determine which repo's to query? I see you order the repositories by it but its not clear to me what ordering is used for. Are there alternatives that could work such as passing an array of repositories (names or ids) to the django command?

**Comment by sarah-clements** [comment]
@Archaeopteryx I'll be on PTO through Aug 2nd, so please ask bhearsum and/or jmaher for a review when you've made some changes and addressed questions.

File: tests/model/test_files_bugzilla_map.py (hunk 3 above)
Comment (Archaeopteryx) [reply tests/model/test_files_bugzilla_map.py:25]: `life_cycle_order` determines the priority of the repository for the file to Bugzilla product+component mapping. A repository with a low value gets processed first and data for the same file path in repositories processed later will be ignored. This allows to require Bugzilla component reorganizations to only require changes on mozilla-central. The other sheriffed repositories need to be processed to have the data for files which got moved or deleted in mozilla-central.

File: treeherder/model/models.py (hunk 2 above)
Comment (Archaeopteryx) [reply treeherder/model/models.py:360]: A current run creates 326k rows. When mozilla-esr78 switches to inactive in ~3 months, it will drop to 292k. The expiration is part of the ingestion script (`paths_removed`). The manual action needed is to remove the `life_cycle_order` attribute from `repository.json` when ESR trees reach their EOL.

File: treeherder/webapp/api/bug_creation.py (hunk 1 above)
Comment (Archaeopteryx) [reply treeherder/webapp/api/bug_creation.py:50]: For context: This feature gets used 20-50 times per day, almost exclusively by code sheriffs.
* The tests which are not web-platform-tests will almost all be covered by the first query.
* The second query is more an edge case.
* The big share of web-platform-test failures will have results after query 3.
* The web-platform-tests which use a generated file name need query #4.
The `path` and `file_name` columns are indexed.

Regarding the use of a custom manager:
The rule `.filter(file_name__startswith=file_without_extension)` from the last query could be used to define the base queryset for the manager but the other filtering rules of query 1-3 would still need to be applied in custom methods. models.py's `FailuresQuerySet` executes the query each time from the full row set (by calling `self.<method>`) if I understand it correctly.

A request to the API endpoint is rather quick based on local testing, e.g. faster than the generation of the bugs suggested for failure lines if there are several failure lines, even with empty `bugscache`.

Can we land the current code for the bugzilla product+component endpoint like it is and evaluate if further work is needed based on performance in production? If the code should be written or organized different, could you provide further details, please?

File: docker/entrypoint_prod.sh [hunk 4]
```diff
@@ -76,6 +76,9 @@ elif [ "$1" == "run_intermittents_commenter" ]; then
 elif [ "$1" == "update_bugscache" ]; then
     newrelic-admin run-program ./manage.py update_bugscache
 
+elif [ "$1" == "update_files_bugzilla_map" ]; then
+    newrelic-admin run-program ./manage.py update_files_bugzilla_map
```
Comment (Archaeopteryx) [inline docker/entrypoint_prod.sh:80]: Where are the cron tasks and their intervals set up? Heroku had a viewer, I didn't find them in GCE.

File: treeherder/webapp/api/bug_creation.py (hunk 1 above)
Comment (sarah-clements) [reply treeherder/webapp/api/bug_creation.py:50]: > The rule .filter(file_name__startswith=file_without_extension) from the last query could be used to define the base queryset for the manager but the other filtering rules of query 1-3 would still need to be applied in custom methods. models.py's FailuresQuerySet executes the query each time from the full row set (by calling self.<method>) if I understand it correctly

I was trying to figure out how to make the code a little more DRY, but maybe its not worth the effort. Do you feel confident trying this out on stage or do you want to test it on prototype branch first?

File: tests/model/test_files_bugzilla_map.py (hunk 3 above)
Comment (sarah-clements) [reply tests/model/test_files_bugzilla_map.py:25]: ok, thanks for the explanation.

File: docker/entrypoint_prod.sh (hunk 4 above)
Comment (sarah-clements) [reply docker/entrypoint_prod.sh:80]: That's something cloudOps will do: https://treeherder.readthedocs.io/infrastructure/administration.html#adding-or-changing-scheduled-tasks-and-environment-variables

File: treeherder/webapp/api/bug_creation.py (hunk 1 above)
Comment (Archaeopteryx) [reply treeherder/webapp/api/bug_creation.py:50]: Having it run on prototype and checking the logs the import succeeds there would be beneficial. After the recent docker upgrade, 2 imports had failed locally because the docker container crashed. Memory was set to 2 GB.

Comment (sarah-clements) [reply treeherder/webapp/api/bug_creation.py:50]: Ok, so you'll want to push your branch to the prototype branch in this repo and your pr should automatically deploy to the prototype instance. Then if you file a bug with cloudOps, they can add set up the cron job. Since oremj is leaving it might not be very quick... no idea who is replacing him as our tech support.

File: docker/entrypoint_prod.sh (hunk 4 above)
Comment (cvalaas) [reply docker/entrypoint_prod.sh:80]: @Archaeopteryx do you have a cron schedule in mind for this?

Comment (Archaeopteryx) [reply docker/entrypoint_prod.sh:80]: Sheriffs merge every 6 or 12 hours at (x * 6 + 4) UTC hours and the task completes soon after. The initial run on this machines is maybe 2 minutes, subsequent ones faster. If you are asking for a full hour: 4 times a day at 5, 11, 17, 23 UTC.

**Comment by Archaeopteryx** [comment]
This looks good on prototype.

**Review by sarah-clements — APPROVED** [review]
